system/lib/wasmfs/paths.cpp

```diff
@@ -46,12 +75,32 @@ ParsedPath getParsedPath(std::vector<std::string> pathParts,
     }
   }
 
-  for (auto pathPart = begin; pathPart != pathParts.end() - 1; ++pathPart) {
+  for (auto pathPart = begin; pathPart != pathParts.end(); ++pathPart) {
     // Find the next entry in the current directory entry
 #ifdef WASMFS_DEBUG
     curr->locked().printKeys();
 #endif
-    auto child = curr->locked().getChild(*pathPart);
+
+    auto lockedCurr = curr->locked();
+
+    // Find the relevant child for this path part.
+    std::shared_ptr<File> child;
+    if (*pathPart == ".") {
+      child = lockedCurr.unlocked();
+    } else if (*pathPart == "..") {
+      if (curr != wasmFS.getRootDirectory()) {
+        child = lockedCurr.getParent()->cast<Directory>();
+      } else {
+        child = lockedCurr.unlocked();
+      }
```

Comment:
I wonder if we should set the root directory as its own parent to avoid special casing it like this in multiple places.

Reply:
That's a good question. It would simplify code, but it is technically "wrong" for a node to be its own parent... It would also mean that we need to carefully make that change when making something the root (say as a result of a move). Not sure I prefer that.

Reply:
I'm not sure it is wrong in the case of root. This logic already makes root its own parent for all intents and purposes, and I believe that in a real file system, the `..` directory entry of root really does point to itself.

Also, it's impossible to rename something to root because rename can only overwrite empty directories. And if root is empty, then there's nothing that can be renamed. I don't believe there's any way to overwrite or replace the root at all.

The only thing we'd need to be careful of would be manually undoing the shared_ptr loop in the WasmFS destructor so that all the in-memory File objects get destructed.